lib/Sema/CSStep.cpp

```diff
@@ -525,29 +528,132 @@ StepResult DisjunctionStep::resume(bool prevFailed) {
+
+  auto genericSignatureA = typeA->getGenericSignature();
+  auto genericSignatureB = typeB->getGenericSignature();
+
+  // Substitute generic parameters with their archetypes in each generic function.
+  Type substTypeA = typeA->substGenericArgs(
+      genericSignatureA->getGenericEnvironment()->getForwardingSubstitutionMap());
+  Type substTypeB = typeB->substGenericArgs(
+      genericSignatureB->getGenericEnvironment()->getForwardingSubstitutionMap());
+
+  // Attempt to substitute archetypes from the second type with archetypes in the
+  // same structural position in the first type.
+  TypeSubstitutionMap substMap;
+  substTypeB = substTypeB->substituteBindingsTo(substTypeA,
+      [&](ArchetypeType *origType, CanType substType,
+          ArchetypeType *, ArrayRef<ProtocolConformanceRef>) -> CanType {
+    auto interfaceTy = origType->getInterfaceType()->getCanonicalType();
+    substMap[interfaceTy->getAs<SubstitutableType>()] = substType;
+    return substType;
+  });
+
+  if (!substTypeB)
+    return false;
+
+  auto result = TypeChecker::checkGenericArguments(
+      declA->getDeclContext(), SourceLoc(), SourceLoc(), typeB,
+      genericSignatureB->getGenericParams(),
+      genericSignatureB->getRequirements(),
+      QueryTypeSubstitutionMap{ substMap });
+
+  if (result != RequirementCheckResult::Success)
+    return false;
+
+  return substTypeA->isEqual(substTypeB);
+}
+
+static bool isDeclSubstitutable(ValueDecl *declA, ValueDecl *declB) {
+  return evaluateOrDefault(declA->getASTContext().evaluator,
+                           IsDeclSubstitutableRequest{ declA, declB },
+                           false);
+}
+
 bool DisjunctionStep::shouldSkip(const DisjunctionChoice &choice) const {
   auto &ctx = CS.getASTContext();
 
-  bool attemptFixes = CS.shouldAttemptFixes();
-  // Enable all disabled choices in "diagnostic" mode.
-  if (!attemptFixes && choice.isDisabled()) {
+  // Never skip disjunction choices in diagnostic mode.
+  if (CS.shouldAttemptFixes())
+    return false;
+
+  auto skip = [&](std::string reason) -> bool {
     if (CS.isDebugMode()) {
       auto &log = getDebugLogger();
-      log << "(skipping ";
+      log << "(skipping " + reason + " ";
       choice.print(log, &ctx.SourceMgr);
       log << '\n';
     }
 
     return true;
-  }
+  };
 
-  // Skip unavailable overloads unless solver is in the "diagnostic" mode.
-  if (!attemptFixes && choice.isUnavailable())
-    return true;
+  if (choice.isDisabled())
+    return skip("disabled");
+
+  // Skip unavailable overloads.
+  if (choice.isUnavailable())
+    return skip("unavailable");
 
   if (ctx.TypeCheckerOpts.DisableConstraintSolverPerformanceHacks)
     return false;
 
+  // If the solver already found a solution with a better overload choice that
+  // can be unconditionally substituted by the current choice, skip the current
+  // choice.
+  if (LastSolvedChoice && LastSolvedChoice->second == getCurrentScore() &&
+      choice.isGenericOperator()) {
```

Comment:
I like the idea behind this, maybe it could be improved by doing this filtering only once upon encountering a new disjunction since `TypeChecker::checkGenericArguments` seems like an expensive operation to repeat and substitution always means global best?

Reply:
Yes, I'm working on making the solver check this less often, but the nice thing about this request is that it's independent of what's being type checked, so it can be cached and reused again across constraint systems. This is not true of `compareDeclarations`, so I am just not going to use it in these heuristics. Also, the substitution is guaranteed to produce the same function type so in theory it should not affect overload resolution elsewhere in the expression (though for this to be 100% true I think I need to change the request to make sure the argument and result types are plain type parameters)

Reply:
bq. it can be cached and reused again across constraint systems

That makes sense but my point is more of a logical place to do that check, I think it makes sense to filter on arrival since this is a stable operation even if it does cache the result.

Reply:
Ah I see, the issue is that we can only filter out more general overloads if the solver finds a solution with the more refined one.

A more sophisticated version of this that @DougGregor pointed out is we can identify overloads that are pure refinements of a more general overload, and make the solver only try the most general one. Then, the solver can select a more refined overload at the end once it has found a solution with the most general one. I think that's out of scope for this PR, but a super interesting idea that we should think through.